src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java

```diff
@@ -791,24 +774,20 @@ private double calculateScore(final ProgrammingExercise programmingExercise, fin
      *
      * @param programmingExercise which the result belongs to.
      * @param result for which the points should be calculated.
-     * @param successfulTestCases all test cases the submission passed.
-     * @param totalTestCaseCount the total number of relevant test cases. This might not be the total
-     *                           number of test cases in the exercise as some test cases are ignored
-     *                           for the calculation before the exercise due date.
-     * @param weightSum the sum of test case weights of all test cases that have to be considered.
      * @return the total score for this result without penalty deductions.
      */
-    private double calculateSuccessfulTestPoints(final ProgrammingExercise programmingExercise, final Result result, final Set<ProgrammingExerciseTestCase> successfulTestCases,
-            int totalTestCaseCount, double weightSum) {
-        double successfulTestPoints = successfulTestCases.stream().mapToDouble(test -> {
-            double credits = calculatePointsForTestCase(result, programmingExercise, test, totalTestCaseCount, weightSum);
-            setCreditsForTestCaseFeedback(result, test, credits);
-            return credits;
-        }).sum();
+    private double calculateSuccessfulTestPoints(final ProgrammingExercise programmingExercise, final Result result, final Set<ProgrammingExerciseTestCase> allTests) {
+        Set<ProgrammingExerciseTestCase> successfulTestCases = allTests.stream().filter(testCase -> testCase.isSuccessful(result)).collect(Collectors.toSet());
+        double successfulTestPoints = successfulTestCases.stream()
+                .mapToDouble(test -> calculatePointsForTestCase(result.getParticipation(), programmingExercise.getMaxPoints(), test, allTests)).sum();
 
         return capPointsAtMaximum(programmingExercise, successfulTestPoints);
     }
 
+    private double calculateWeightSum(final Set<ProgrammingExerciseTestCase> allTests) {
```

Comment:
This is now called for each test case from `calculatePointsForTestCase` and iterating over all test cases again instead of only once, even though it does not change.

Reply:
Yes, the runtime is a little bit worse but I think for now this is preferable to having 5 parameters in that function.

Reply:
I'd also prefer one more parameter than calling methods unnecessarily often.

Reply:
There are two reasons for having that here:
1. One less parameter, leading to easier reuse of that method
2. Clearer separation of concerns. We used to set feedback.credits in `calculateSuccesfulTestPoints`. i.e. we had a method that not only calculated but also set the feedback credits. This method was also called through this call stack: `calculateScoreForResult -> calculateScoreForResult -> updateScore -> calculateScore -> calculateSuccesfulTestPoints`, none of which mentioning feedbacks and leading to more mental overhead understanding the code. I think having a runtime of `O(2n) compared to O(n)` as a transient state for future refactoring and clearer separation of concerns is a worthy trade-off

Reply:
> Clearer separation of concerns. We used to set feedback.credits in calculateSuccesfulTestPoints. i.e. we had a method that not only calculated but also set the feedback credits. This method was also called through this call stack: calculateScoreForResult -> calculateScoreForResult -> updateScore -> calculateScore -> calculateSuccesfulTestPoints, none of which mentioning feedbacks and leading to more mental overhead understanding the code.

Yes, it definitely should remain its own method and the `calculateSuccesfulTestPoints` should not set the feedback credits. I agree on that.


> runtime of `O(2n)`

Is it? For each test the sum is calculated by looping over all tests again → O(n²)


> One less parameter, leading to easier reuse of that method

What about creating a private record that contains both all tests and their weight sum? The you could save the additional parameter and it is ensured that the tests and their weights are always together making sure that not the wrong list of tests is used to calculate the sum.
Or to make it more general if that makes sense in more places, include the exercise and/or participation as well to contain the helper data that is not changed but required to calculate the points.

Reply:
yeah you're right with the runtime. I thought I called that differently.

I added a private record. Maybe you could have a look?